navigation/perfect_memory_base.h

```diff
@@ -57,65 +57,65 @@ class PerfectMemoryBase
+    /*!
+     * \brief Get differences between current view and stored snapshots
+     *
+     * The parameters are perfect-forwarded to the Rotater class, so e.g. for
+     * InSilicoRotater one passes in a cv::Mat and (optionally) an unsigned int
+     * for the scan step and for the AntWorldRotater, one passes in one or more
+     * angles.
+     */
+    template<class... Ts>
+    std::vector<std::vector<float>> getImageDifferences(Ts &&... args) const
     {
-        const auto &unwrapRes = getUnwrapResolution();
-        assert(image.cols == unwrapRes.width);
-        assert(image.rows == unwrapRes.height);
-        assert(image.type() == CV_8UC1);
         const size_t numSnapshots = getNumSnapshots();
         assert(numSnapshots > 0);
-        const size_t scanStep = getScanStep();
 
-        // Clone mask and image so they can be rolled in place
-        getMaskImage().copyTo(m_ScratchMaskImage);
-        image.copyTo(m_ScratchRollImage);
+        Rotater rotater(getUnwrapResolution(), getMaskImage(), std::forward<Ts>(args)...);
 
         // Create vector to store RIDF values
         std::vector<std::vector<float>> differences(numSnapshots);
         for (auto &d : differences) {
-            d.resize(m_ScratchRollImage.cols);
+            d.resize(rotater.max());
         }
 
         // Scan across image columns
-        for (int col = 0; col < m_ScratchRollImage.cols; col += scanStep) {
+        rotater([this, &differences, numSnapshots](const cv::Mat &fr, const cv::Mat &mask, size_t i) {
             // Loop through snapshots
             for (size_t s = 0; s < numSnapshots; s++) {
                 // Calculate difference
-                differences[s][col] = calcSnapshotDifference(m_ScratchRollImage, m_ScratchMaskImage, s);
-            }
-
-            // Roll image and corresponding mask left by scanStep
-            rollImage(m_ScratchRollImage);
-            if (!m_ScratchMaskImage.empty()) {
-                rollImage(m_ScratchMaskImage);
+                const auto diff = calcSnapshotDifference(fr, mask, s);
+                differences[s][i] = diff;
             }
-        }
+        });
 
         return differences;
     }
 
-    //! Get an estimate for heading based on comparing image with stored snapshots
-    auto getHeading(const cv::Mat &image) const
+    /*!
+     * \brief Get an estimate for heading based on comparing image with stored
+     *        snapshots
+     *
+     * The parameters are perfect-forwarded to the Rotater class, so e.g. for
+     * InSilicoRotater one passes in a cv::Mat and (optionally) an unsigned int
+     * for the scan step and for the AntWorldRotater, one passes in one or more
+     * angles.
+     */
+    template<class... Ts>
+    auto getHeading(Ts &&... args) const
     {
-        std::vector<std::vector<float>> differences = getImageDifferences(image);
+        auto differences = getImageDifferences(std::forward<Ts>(args)...);
```

Comment:
allocating differences (potentially every frame) seems wasteful both in terms of memory (lots of snapshots * hundreds of columns * sizeof(float) isn't nothing and this is a lot of heap allocations) when all you're doing is finding the minimum - would it be possible (at the expense of some extra typing) to repeat the algorithm in this function and calculate the minimum inline?

Reply:
You're right that it probs is a bit expensive. At least we should probably be reusing the vector if nothing else.

But the reason I went with storing the differences in vectors is so that we can get these values back out again for debugging and looking at the internal state. I did benchmark it and performance was basically the same as not having these big heap allocations, but I was only comparing a single frame. We might be able to get both options out by making a little iterator class for these differences though... I'll think about it.

Maybe in the meantime we could just settle for reusing the vector to save on the heap allocations though?

Reply:
Yeah - caching that vector or vectors as a member of the class would indeed be an excellent middle ground.